bootstrap/cmd/bootstrap/app/ksServer.go

```diff
@@ -652,16 +682,20 @@ func GetRepoName(project string) string {
 	return fmt.Sprintf("%s-kubeflow-config", project)
 }
 
+func generateRandStr(length int) string {
+	chars := "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ"
+	b := make([]byte, length)
+	for i := range b {
+		b[i] = chars[rand.Intn(len(chars))]
+	}
+	return string(b)
+}
+
 // Not thread-safe, make sure project lock is on.
 // Clone project repo to local disk, which contains all existing ks apps config in the project
-func (s *ksServer) CloneRepoToLocal(project string, token string) error {
-	repoDir := path.Join(s.appsDir, GetRepoName(project))
-	_, err := s.fs.Stat(repoDir)
-
-	// delete existing local repo, pull from cloud
-	if err == nil {
-		os.RemoveAll(repoDir)
-	}
+func (s *ksServer) CloneRepoToLocal(project string, token string) (string, error) {
```

Comment:
Why isn't this thread safe?

Reply:
The comment is not accurate: it's only not thread-safe when its part of createApp.
createApp need to write to source repo, concurrent push might cause conflict on git side.

Reply:
So createApp action in general is not thread safe when multi user using same project;
createApp currently protected by a project-lock